lightning/src/ln/peers/handshake/mod.rs

```diff
@@ -0,0 +1,405 @@
+				}
+
+				let mut act_one_buffer = [0u8; ACT_ONE_LENGTH];
+				act_one_buffer.copy_from_slice(&self.read_buffer[..ACT_ONE_LENGTH]);
+				self.read_buffer.drain(..ACT_ONE_LENGTH);
+
+				let act_two = self.process_act_one(ActOne(act_one_buffer))?;
+				response = Some(Act::Two(act_two));
+			}
+			&Some(HandshakeState::AwaitingActTwo(_)) => {
+				if read_buffer_length < ACT_TWO_LENGTH {
+					return Err("need at least 50 bytes".to_string());
+				}
+
+				let mut act_two_buffer = [0u8; ACT_TWO_LENGTH];
+				act_two_buffer.copy_from_slice(&self.read_buffer[..ACT_TWO_LENGTH]);
+				self.read_buffer.drain(..ACT_TWO_LENGTH);
+
+				let (act_three, mut conduit) = self.process_act_two(ActTwo(act_two_buffer))?;
+
+				if self.read_buffer.len() > 0 { // have we received more data still?
+					conduit.read(&self.read_buffer[..]);
+					self.read_buffer.drain(..);
+				}
+
+				response = Some(Act::Three(act_three));
+				connected_peer = Some(conduit);
+			}
+			&Some(HandshakeState::AwaitingActThree(_)) => {
+				if read_buffer_length < ACT_THREE_LENGTH {
+					return Err("need at least 66 bytes".to_string());
+				}
+
+				let mut act_three_buffer = [0u8; ACT_THREE_LENGTH];
+				act_three_buffer.copy_from_slice(&self.read_buffer[..ACT_THREE_LENGTH]);
+				self.read_buffer.drain(..ACT_THREE_LENGTH);
+
+				let (public_key, mut conduit) = self.process_act_three(ActThree(act_three_buffer))?;
+
+				if self.read_buffer.len() > 0 { // have we received more data still?
+					conduit.read(&self.read_buffer[..]);
+					self.read_buffer.drain(..);
+				}
+
+				connected_peer = Some(conduit);
+				self.remote_public_key = Some(public_key);
+			}
+			_ => {
+				return Err("no acts left to process".to_string());
+			}
+		};
+		Ok((response, connected_peer))
+	}
+
+	/// Initiate the handshake with a peer and return the first act
+	pub fn initiate(&mut self, remote_public_key: &PublicKey) -> Result<ActOne, String> {
+		if let &Some(HandshakeState::Uninitiated) = &self.state {} else {
+			return Err("Handshakes can only be initiated from the uninitiated state".to_string());
+		}
+
+		let (mut hash, chaining_key) = Self::initialize_state(&remote_public_key);
+
+		// serialize act one
+		let (act_one, chaining_key, temporary_key) = Self::calculate_act_message(
+			&self.ephemeral_private_key,
+			remote_public_key,
+			chaining_key,
+			&mut hash,
+		);
+
+		self.state = Some(HandshakeState::AwaitingActTwo(ActTwoExpectation {
+			hash,
+			chaining_key,
+			temporary_key,
+			ephemeral_private_key: (*&self.ephemeral_private_key).clone(),
+		}));
+
+		Ok(ActOne(act_one))
+	}
+
+	/// Process a peer's incoming first act and return the second act
+	pub(crate) fn process_act_one(&mut self, act: ActOne) -> Result<ActTwo, String> {
+		let state = self.state.take();
+		let act_one_expectation = match state {
+			Some(HandshakeState::AwaitingActOne(act_state)) => act_state,
+			Some(HandshakeState::Uninitiated) => {
+				// this can also be initiated from a blank state
```

Comment:
This comment can be left out IMHO. I think it will confuse the casual reader more than anything. "blank" is referring to the old name of "HandshakeState::Uninitiated" so without that context seems like this comment is referring to `private_key_to_public_key`.